blogs/treewidth.md

```diff
@@ -0,0 +1,385 @@
+
+![](/assets/treewidth_figs/treesa.png)
+
+The TreeSA method starts from a random contraction tree and then applies the above rules to transform the tree. The cost of the contraction tree is evaluated and the tree is updated according to the Metropolis criterion. During the process, the temperature is gradually decreased, and the process stop when the temperature is low enough.
+The method has already been used in OMEinsumContractionOrders.jl.
+
+
+The method listed above are powerful and efficient, which can be used to find great contraction orders of the extremely large tensor networks.
+However, they are heuristic methods and may not guarantee to find the optimal contraction order. In the following sections, we will introduce a theoretical method to find the optimal contraction order based on the tree decomposition.
+
+
+## Finding the Optimal Contraction Order
+
+In the previous section, we introduce the concept of tensor network and its contraction order, so that now you should understand why the contraction order so important. 
+Then the next question is how to find the optimal contraction order.
+
+In this section, we will introduce one way to find the theoretical optimal contraction order based on the exact tree width solver, according to the following well known theorem[^Markov] :
+
+**Theorem 1**. Let $C$ be a quantum circuit with $T$ gates and whose underlying circuit graph is $G_c$. Then $C$ can be simulated deterministically in time $T^{O(1)} e^{O(tw(G_C))}$, where $tw(G_C)$ is the treewidth of $G_C$.
+
+Using the language of tensor network, we can rewrite the above theorem as follows: the bottleneck of time complexity of the contraction of a tensor network is $O(e^{O(tw(L(G)))})$, where $L(G)$ is the line graph of the hypergraph of the tensor network. Therefore, if we can find the tree decomposition of the tensor network with minimal treewidth, we can find the optimal contraction order of the tensor network.
+
+To further make use of the theorem, we need to introduce the concepts of line graph, tree width and tree decomposition.
+
+### Line Graph
+
+A formal definition of the [**line graph**](https://en.wikipedia.org/wiki/Line_graph) is as follows:
+
+**Definition 1**. Given a graph G, its line graph $L(G)$ is a graph such that: 
+   1. each vertex of $L(G)$ represents an edge of $G$; 
+   2. two vertices of $L(G)$ are adjacent if and only if their corresponding edges share a common endpoint ("are incident") in $G$. 
+
+That is, it is the intersection graph of the edges of G, representing each edge by the set of its two endpoints.
+
+For a tensor network, we can construct a hypergraph $G$ whose vertices are the tensors and whose hyperedges are the indices. Then the line graph $L(G)$ of the hypergraph $G$ is the graph whose vertices are the indices and whose edges are the tensors, which is a simple graph, as shown in the following figure:
```

Comment:
Is there a simple way to reconstruct the original graph/tensor network contraction from the line graph?

Reply:
I think it hard to do so, since different graphs can corresponding the same line graph if we consider hyper graphs. An example is given below:
<img width="688" alt="image" src="https://github.com/user-attachments/assets/1880f0dd-b9b8-4d75-846a-ea905d7018bc">